### PR TITLE
[7.x] Add HttpClient event

### PR DESCRIPTION
**Pull request by domthomas-dev**
This PR allows to add HttpClient Event named PendingRequestSent.

We can have url, method,options and response of this Request.

### CODE REVIEW

File: src/Illuminate/Http/Client/Factory.php [hunk 1]
```diff
@@ -42,15 +43,20 @@ class Factory
      * @var array
      */
     protected $responseSequences = [];
+    /**
+     * @var Dispatcher|null
```
Comment (GrahamCampbell) [inline src/Illuminate/Http/Client/Factory.php:47]: Please use fully qualified class names in phpdoc.

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

File: src/Illuminate/Http/Client/Factory.php (hunk 1 above)
Comment (thomasdominic) [reply src/Illuminate/Http/Client/Factory.php]: It's okay, and I use it in the PendingRequestEvent File too

File: src/Illuminate/Http/Client/Event/PendingRequestSent.php [hunk 2]
```diff
@@ -16,16 +16,20 @@ class PendingRequestSent
     /**
      * The Http client Response.
      *
-     * @var Response|null
+     * @var Illuminate\Http\Client\Response|null
```
Comment (GrahamCampbell) [inline src/Illuminate/Http/Client/Event/PendingRequestSent.php:19]: Missing leading slash.

File: src/Illuminate/Http/Client/Factory.php [hunk 3]
```diff
@@ -44,7 +44,7 @@ class Factory
      */
     protected $responseSequences = [];
     /**
-     * @var Dispatcher|null
+     * @var Illuminate\Contracts\Events\Dispatcher|null
```
Comment (GrahamCampbell) [inline src/Illuminate/Http/Client/Factory.php:47]: same here ;)

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

File: src/Illuminate/Http/Client/Factory.php (hunk 3 above)
Comment (thomasdominic) [reply src/Illuminate/Http/Client/Factory.php]: Sorry, it's done

File: src/Illuminate/Http/Client/Event/PendingRequestSent.php [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+     */
+    public $options;
+
+    /**
+     * Create a new event instance.
+     *
+     * @param string $method
+     * @param string $url
+     * @param array $options
+     * @param \Illuminate\Http\Client\Response $response
```
Comment (GrahamCampbell) [inline src/Illuminate/Http/Client/Event/PendingRequestSent.php:43]: Missing `|null`

File: src/Illuminate/Http/Client/Factory.php [hunk 5]
```diff
@@ -42,15 +43,20 @@ class Factory
...
     protected $responseSequences = [];
+    /**
+     * @var \Illuminate\Contracts\Events\Dispatcher|null
+     */
+    private $dispatcher;
 
     /**
      * Create a new factory instance.
      *
-     * @return void
```
Comment (GrahamCampbell) [inline src/Illuminate/Http/Client/Factory.php:49]: please keep this

File: src/Illuminate/Http/Client/PendingRequest.php [hunk 6]
```diff
@@ -108,10 +117,10 @@ class PendingRequest
     /**
      * Create a new HTTP Client instance.
      *
-     * @param  \Illuminate\Http\Client\Factory|null  $factory
-     * @return void
```
Comment (GrahamCampbell) [inline src/Illuminate/Http/Client/PendingRequest.php:112]: Please keep this

File: src/Illuminate/Http/Client/PendingRequest.php [hunk 7]
```diff
@@ -108,10 +117,10 @@ class PendingRequest
     /**
      * Create a new HTTP Client instance.
      *
-     * @param  \Illuminate\Http\Client\Factory|null  $factory
```
Comment (GrahamCampbell) [inline src/Illuminate/Http/Client/PendingRequest.php:111]: Please don't delete these spaces, and make sure your other param doc also has 2 spaces either side of the type.

File: src/Illuminate/Http/Client/PendingRequest.php [hunk 8]
```diff
@@ -731,4 +743,22 @@ public function stub($callback)
 
         return $this;
     }
+
+    /**
+     * Dispatch the Pending Request sent event.
+     *
+     * @param $method
```
Comment (GrahamCampbell) [inline src/Illuminate/Http/Client/PendingRequest.php:750]: invalid syntax

File: src/Illuminate/Http/Client/PendingRequest.php [hunk 9]
```diff
@@ -731,4 +743,22 @@ public function stub($callback)
...
+     * @param $url
```
Comment (GrahamCampbell) [inline src/Illuminate/Http/Client/PendingRequest.php:751]: invalid syntax

File: src/Illuminate/Http/Client/PendingRequest.php [hunk 10]
```diff
@@ -731,4 +743,22 @@ public function stub($callback)
...
+     * @param $options
```
Comment (GrahamCampbell) [inline src/Illuminate/Http/Client/PendingRequest.php:752]: invalid syntax

File: src/Illuminate/Http/Client/PendingRequest.php [hunk 11]
```diff
@@ -731,4 +743,22 @@ public function stub($callback)
...
+     * @param null $response
```
Comment (GrahamCampbell) [inline src/Illuminate/Http/Client/PendingRequest.php:753]: incorrect type

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

**Comment by thomasdominic** [comment]
Hi @GrahamCampbell All are done, and sorry for that.

File: src/Illuminate/Http/Client/PendingRequest.php (hunk 11 above)
Comment (thomasdominic) [reply src/Illuminate/Http/Client/PendingRequest.php]: Done

**Comment by driesvints** [comment]
@thomasdominic can you squash your commits?

**Comment by taylorotwell** [comment]
Thanks for your pull request to Laravel!

Unfortunately, I'm going to delay merging this code for now. To preserve our ability to adequately maintain the framework, we need to be very careful regarding the amount of code we include. 

If possible, please consider releasing your code as a package so that the community can still take advantage of your contributions! 

If you feel absolutely certain that this code corrects a bug in the framework, please "**@**" mention me in a follow-up comment with further explanation so that GitHub will send me a notification of your response.

**Comment by thomasdominic** [comment]
Hi, Thanks for your reply @taylorotwell
If you look at my PR, you can see that there is little code and there are more comments and PhpDoc.

All the great features of Laravel allow you to have event triggers, HTTPClient is one of the few not to have it.

The code is: an event basic class, calling it in a method and including the dispatcher. I did Unit test, and older HttpClient test are okay.
Could you review your decision? I don't even see how it would be possible with a package 
Or tagged it for Laravel 8?

**Pull request closed without merging**